Cython/Utility/CConvert.pyx

```diff
@@ -130,3 +130,69 @@ cdef inline tuple {{to_tuple_cname}}({{base_type}} *v, Py_ssize_t length):
         Py_INCREF(value)
         PyTuple_SET_ITEM(t, i, value)
     return t
+
+#################### c_array_slice.to_py ####################
+
+cdef extern from *:
+    void Py_INCREF(object o)
+    object PyErr_Format(exc, const char *format, ...)
+    int __Pyx_is_valid_index(Py_ssize_t i, Py_ssize_t limit)
+
+    list PyList_New(Py_ssize_t size)
+    void PyList_SET_ITEM(object p, Py_ssize_t pos, object o)
+
+    tuple PyTuple_New(Py_ssize_t size)
+    void PyTuple_SET_ITEM(object p, Py_ssize_t pos, object o)
+
+    set PySet_New(void *)
+    void PySet_Add(object p, object o)
+
+    set PyFrozenSet_New(void *)
+
+
+@cname("{{cname}}")
+cdef inline {{to_py_obj_type}} {{cname}}({{base_type}} *arr, Py_ssize_t arr_size, Py_ssize_t from_idx, Py_ssize_t to_idx, int wraparound, int boundscheck):
+    if wraparound:
+        if from_idx < 0: 
+            from_idx += arr_size
+
+        if to_idx < 0: 
+            to_idx += arr_size
+
+    if boundscheck and\
+         not (__Pyx_is_valid_index(from_idx, arr_size) and __Pyx_is_valid_index(to_idx - 1, arr_size)):
+            PyErr_Format(IndexError, "Some of slice indexes is out of range")
+
+
+    cdef Py_ssize_t i, length = to_idx - from_idx
+
+    if boundscheck and length < 0:
+        PyErr_Format(IndexError, "Wrong range of slice")
+
+    {{if to_py_obj_type == 'list'}}
+    l = PyList_New(length)
+    {{elif to_py_obj_type == 'tuple' or to_py_obj_type == 'frozenset'}}
+    l = PyTuple_New(length)
+    {{elif to_py_obj_type == 'set'}}
+    l = PySet_New(NULL)
+    {{endif}}
+
+    cdef object value
+    for i in range(<size_t> from_idx, <size_t> to_idx):
+
+        value = arr[i]
+        Py_INCREF(value)
+
+        {{if to_py_obj_type == 'list'}}
+        PyList_SET_ITEM(l, i - from_idx, value)
+        {{elif to_py_obj_type == 'tuple' or to_py_obj_type == 'frozenset'}}
+        PyTuple_SET_ITEM(l, i - from_idx, value)
+        {{elif to_py_obj_type == 'set'}}
+        PySet_Add(l, value)
```

Comment:
Unlike `PyList_SET_ITEM` and `PyTuple_SET_ITEM` this can raise an exception. It also doesn't steal a reference, so I think you have a memory leak.

Reply:
ok, will fix